### ISSUE TITLE
How to retry streaming uploads

### ISSUE THREAD

**Issue by theyoprst** [kennethreitz/requests#3905]
Consider this piece of code:

```python
retries = Retry(total=3)
session = requests.Session()
session.mount('http://', HTTPAdapter(max_retries=retries))
with open('big_file.txt', 'rb') as f:
    session.put(url, data=f)
```

It has a bug.
When network error occurred during transmission of the file, the corresponding file handle points to some position _inside_ the file. So the next retry performs new HTTP connection which uploads _the rest_ of the file. Thankfully, in this case, `Content-Length` doesn't change, so the server waits for the rest of data and then disconnects after some timeout.
Here is [the script](https://gist.github.com/theyoprst/df536545e1a4a0f3f21cf584a547a529) which demonstrates this bug. And here is [it's output](https://gist.github.com/theyoprst/7c69d019457d89310f16707e979c1580).

So retries are almost useless for such uploads (they _are_ useful in case of connection problems, though).
Now I can only think of external retrying by manual reopening / re-seeking of the file.
Is there any way to make requests to retry it 'right'?
Should this behavior be fixed in requests?

**Comment by theyoprst** [comment]
Wow. It seems in the latest version it was already fixed.
Probably here: https://github.com/shazow/urllib3/issues/459

**Comment by nateprewitt** [comment]
Hey @theyoprst, you're correct, this should have been resolved in urllib3 in shazow/urllib3#1039. Are you still experiencing this issue in the latest version of requests?

**Comment by nateprewitt** [comment]
Yep, as it looks like you were referring to the latest version of Requests. This doesn't seem to be reproducible with 2.13.0 which has the patch included in urllib3. Thanks for taking the time to open this issue, @theyoprst!
